src/profile-logic/process-profile.js

```diff
@@ -1059,53 +1066,193 @@ export function processProfile(
  */
-export function serializeProfile(
-  profile: Profile,
-  includeNetworkUrls: boolean = true
-): string {
+export function serializeProfile(profile: Profile): string {
   // stringTable -> stringArray
+  const newProfile = Object.assign({}, profile, {
+    threads: profile.threads.map(thread => {
+      const stringArray = thread.stringTable.serializeToArray();
+      // Has to be any since Threads don't have stringArray.
+      const newThread: any = Object.assign({}, thread);
+      delete newThread.stringTable;
+      newThread.stringArray = stringArray;
+      return newThread;
+    }),
+  });
+
+  return JSON.stringify(newProfile);
+}
+
+/**
+ * Take a processed profile with PII that user wants to be removed and remove the
+ * thread data depending on that PII status. Look at `RemoveProfileInformation`
+ * type definition if you want to learn what kind of information we are removing.
+ */
+export function sanitizePII(
+  profile: Profile,
+  PIIToBeRemoved: RemoveProfileInformation
+): Profile {
   let urlCounter = 0;
+  const removedThreadIndexes = [];
   const newProfile = Object.assign({}, profile, {
-    meta: { ...profile.meta, networkURLsRemoved: !includeNetworkUrls },
+    meta: {
+      ...profile.meta,
+      // FIXME: We won't be needing this after PII frontend changes.
+      networkURLsRemoved: PIIToBeRemoved.shouldRemoveNetworkUrls,
+      extensions: PIIToBeRemoved.shouldRemoveExtensions
+        ? getEmptyExtensions()
+        : profile.meta.extensions,
+    },
     pages:
-      includeNetworkUrls === false && profile.pages
+      (PIIToBeRemoved.shouldRemoveNetworkUrls ||
+        PIIToBeRemoved.shouldRemoveAllUrls) &&
+      profile.pages
         ? profile.pages.map(page =>
             Object.assign({}, page, {
               url: 'Page #' + urlCounter++,
             })
           )
         : profile.pages,
-    threads: profile.threads.map(thread => {
-      const stringArray = thread.stringTable.serializeToArray();
-      const newThread = Object.assign({}, thread);
-      delete newThread.stringTable;
-      if (includeNetworkUrls === false) {
-        for (let i = 0; i < newThread.markers.length; i++) {
-          const currentMarker = newThread.markers.data[i];
-          if (
-            currentMarker &&
-            currentMarker.type &&
-            currentMarker.type === 'Network'
-          ) {
-            // Remove the URI fields from marker payload.
-            currentMarker.URI = '';
-            currentMarker.RedirectURI = '';
-            // Strip the URL from the marker name
-            const stringIndex = newThread.markers.name[i];
-            stringArray[stringIndex] = stringArray[stringIndex].replace(
-              /:.*/,
-              ''
-            );
-          }
-        }
+    threads: profile.threads.reduce((acc, thread, threadIndex) => {
+      let newThread: any = Object.assign({}, thread);
+
+      newThread = sanitizeThreadPII(newThread, threadIndex, PIIToBeRemoved);
```

Comment:
It would be nice to make `sanitizeThreadPII` a pure function. Right now it's explicitly mutating the passed in profile.